Enhance `ColorPickerWidget` (#5585)
plone/volto: docs/source/recipes/color-picker-widget.md

---
myst:
  html_meta:
    "description": "How to use the color picker widget in blocks settings and forms"
    "property=og:description": "How to use the color picker widget in blocks settings and forms"
    "property=og:title": "How to use the color picker widget"
    "keywords": "Volto, Plone, frontend, React, blocks, forms, widget, color, picker"
---

# Color picker widget

> Volto comes with a color picker widget that can be used in any Volto form.

Reviewer: Insert a sentence with a link to the Storybook color picker demo.

```suggestion
Volto comes with a color picker widget that can be used in any Volto form.
You can [try a demo of the default color picker](https://6.docs.plone.org/storybook/?path=/story/edit-widgets-colorpicker--default).
```

Author: Ah sorry I forgot.

You can combine the color picker widget with the {doc}`../blocks/block-style-wrapper` to have a powerful, yet simple way to manage color properties in your blocks.
You can use it either in your custom block's setting's schema or enhance an existing block as follows:

> ```js

Reviewer: We need to use reStructuredText fallback syntax, instead of CommonMark shorthand, to get emphasize lines. In a future version of the theme, we will be able to use the shorthand.

````suggestion
```{code-block} js
````

  },
});

export const defaultStylingSchema = ({ schema, formData, intl }) => {
  const BG_COLORS = [
    { name: 'transparent', label: 'Transparent' },
    { name: 'grey', label: 'Grey' },
  ];

  // You could allow passing the color definition from the config or from the default
  // defined above
  const colors =
    config.blocks?.blocksConfig?.[formData['@type']]?.colors || BG_COLORS;

  // Same for the default used (or undefined)
  const defaultBGColor =
    config.blocks?.blocksConfig?.[formData['@type']]?.defaultBGColor;

  // This adds the StyleWrapper support to your block
  addStyling({ schema, intl });

  // Then we add the field to the fieldset inside the StyleWrapper `styles` field schema fieldset array
  schema.properties.styles.schema.fieldsets[0].fields = [
    ...schema.properties.styles.schema.fieldsets[0].fields,
    'backGroundColor',
  ];

  // and finally, we add the field to the StyleWrapper `styles` object field schema properties
  schema.properties.styles.schema.properties['backGroundColor'] = {
    widget: 'color_picker',
    title: intl.formatMessage(messages.backgroundColor),
    colors,
    default: defaultBGColor,
  };

  return schema;
};
```

> The widget name the color picker widget uses is `color_picker`.

Reviewer: Is this better? When you use "uses", it makes it sound like its name can be set to any value, but I don't think that is your intent.

```suggestion
The color picker widget's discriminator is `color_picker`.
```

## Color definitions

> ```{versionchanged} 17.9.0

Reviewer: Extra blank line for author legibility.

````suggestion

```{versionchanged} 17.9.0
````

Enhanced `ColorPickerWidget` with additional color definitions, saving it as an object instead of a string.

> ```

Reviewer: Extra blank line for author legibility.

````suggestion
```

````

---
myst:
  html_meta:
    "description": "How to use the color picker widget in blocks settings and forms"
    "property=og:description": "How to use the color picker widget in blocks settings and forms"
    "property=og:title": "How to use the color picker widget"
    "keywords": "Volto, Plone, frontend, React, blocks, forms, widget, color, picker"
---

# Color picker widget

> Volto comes with a color picker widget that can be used in any Volto form.

Reviewer 2: People might be expecting a color widget that allows picking any color, so we might want to clarify that this one picks from a preset list of colors.

Author: I improved it in the next sentence.

You can [try a demo of the default color picker](https://6.docs.plone.org/storybook/?path=/story/edit-widgets-colorpicker--default).
You can combine the color picker widget with the {doc}`../blocks/block-style-wrapper` to have a powerful, yet simple way to manage color properties in your blocks.

> You can use it either in your custom block's setting's schema or enhance an existing block as follows:

Reviewer 2: ```suggestion
You can use it either in your custom block's styles schema or enhance an existing block as follows:
```

    config.blocks?.blocksConfig?.[formData['@type']]?.colors || BG_COLORS;

  // Same for the default used (or undefined)
  const defaultBGColor =
    config.blocks?.blocksConfig?.[formData['@type']]?.defaultBGColor;

  // This adds the StyleWrapper support to your block
  addStyling({ schema, intl });

  // Then we add the field to the fieldset inside the StyleWrapper `styles` field schema fieldset array
  schema.properties.styles.schema.fieldsets[0].fields = [
    ...schema.properties.styles.schema.fieldsets[0].fields,
    'backGroundColor',
  ];

  // and finally, we add the field to the StyleWrapper `styles` object field schema properties
  schema.properties.styles.schema.properties['backGroundColor'] = {
    widget: 'color_picker',
    title: intl.formatMessage(messages.backgroundColor),
    colors,
    default: defaultBGColor,
  };

  return schema;
};
```

The color picker widget's discriminator is `color_picker`.

## Color definitions

```{versionchanged} 17.9.0
Enhanced `ColorPickerWidget` with additional color definitions, saving it as an object instead of a string.
```

The `colors` property of the widget controls which colors are available to choose in the widget.

> This is the signature of the object:

Reviewer 2: The example below includes both the type signature and an example (which is great, but makes this sentence inaccurate)

```ts
type Color =
  | {
      name: string;
      label: string;
      style: Record<`--${string}`, string>;
    }
  | {
      name: string;
      label: string;
      style: undefined;
    };

const colors: Color[] = [
  {
    name: 'red',
    label: 'red',
    style: { '--background-color': 'red' } },
  {
    name: 'yellow',
    label: 'yellow',
    style: { '--background-color': 'yellow' },
  },
  {
    name: 'green',
    label: 'green'
  },
]
```

### Basic color definition

> The basic color definition is the one that saves a string label as the widget value.

Reviewer 2: Does this come from the `label` or the `name` above?
